[Nam Minh Quan] iP

quannam0124 opened the pull request:
#Duke Chat-bot 
##This chat-bot helps you organize your to-do list. It's
* Easy to use
* Text-base
* Extremely convenient

##All you need to do is:
1.  Download it from [here](https://github.com/quannam0124/ip.git)
2. Double click it
3. Type in your tasks
4. Let it manage your tasks for you 😁 

And it is completely **FREE**!

Features:
- [x] Managing tasks
- [x] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java

 public static void main(String[] args) {
        new Duke("data/Tasks.txt").run();
    }

```

quannam0124 committed:
Add Echo class for Level 1

quannam0124 committed:
Add TaskList class to manage list of tasks
Modify Duke.main() to perform taskList functions

quannam0124 committed:
Add Task class
Refactor TaskList to work with Task objects
Complete Level 3

quannam0124 committed:
Add javadocs to TaskList.java file

quannam0124 committed:
Rename print() method to toString() method in Task.java and TaskList.java

quannam0124 committed:
Add Deadline, Event, and Todo subclasses to Task
Add Input and expected text files for testing
Level 4

quannam0124 committed:
Add Duke Exceptions
Add Exception handlers
Add DukeHandler to handle responses and prevent public access to list of tasks

quannam0124 committed:
Fix bug throws Exceptions even when input is "list"

quannam0124 committed:
Add delete function to TaskList class
Fix handle logic bug
Level 6 done

quannam0124 committed:
Add TaskStorage class

The TaskStorage class handles loading and modifying the text file that saves all tasks

Let's create this class

quannam0124 committed:
Add DateTime converter for Duke

Date and time is stored as String in Task subclasses

Let's make a class that converts string input for Times to LocalDateTime objects to be stored in task objects

quannam0124 committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/DukeHandler.java
#	src/main/java/Event.java

quannam0124 committed:
Refactor to achieve OOP

Duke is using a DukeHandler to parse user input, user interaction is being handled in multiple classes

Let's pull up behaviors common to user interaction and create a new Ui class

This helps OOP to be clearer in this project

quannam0124 committed:
Modify DukeDate class

DukeDate class modified for ease of reading

quannam0124 committed:
Organize classes into packages
A-packages increment completed

quannam0124 committed:
Add Gradle to project

quannam0124 committed:
Add jUnit test cases

Duke does not have test cases to verify its functions

Let's make some jUnit tests to make sure that the functions work

quannam0124 committed:
Modify build.gradle mainClassName to follow current project's name

quannam0124 committed:
Add JavaDoc

quannam0124 committed:
Modify code to follow coding standards

quannam0124 committed:
Add Find function to Duke

Duke does not have function to search for specific tasks

Let's add a findTask() method to allow this

quannam0124 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/Duke/DukeHandler.java
#	src/main/java/Duke/task/TaskList.java

quannam0124 committed:
Merge branch 'branch-Level-9'

quannam0124 committed:
Add javadoc comment for findTask() param in TaskList.java

quannam0124 committed:
Modify Ui method names to follow coding standards

quannam0124 committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Duke.java
#	src/main/java/Event.java

quannam0124 committed:
Add checkstyle

quannam0124 committed:
Add checkstyle to build file

quannam0124 committed:
Changed xml supressions file name to suppressions

quannam0124 committed:
Fix java coding standards violation

Ramanathan0908 commented on README.md:
```diff
@@ -1,4 +1,4 @@
-# Duke project template
+# Duke.Duke project template
```
A very minor nitpick, Duke got refactored to Duke.Duke which does not make sense here.

Ramanathan0908 commented on src/main/java/Duke/Duke.java:
```diff
@@ -0,0 +1,44 @@
+
+import Duke.task.TaskList;
+import Duke.task.TaskStorage;
+import Duke.util.Ui;
```
I would love to see a line separating task and util packages even further.

Ramanathan0908 commented on src/main/java/Duke/task/Task.java:
```diff
@@ -0,0 +1,51 @@
+ * Represents the object that the chat-bot manages
+ * and users can interact with.
+ */
+public abstract class Task {
```
I like how you defined Task as an abstract class 👍

Ramanathan0908 commented on src/main/java/Duke/task/TaskStorage.java:
```diff
@@ -0,0 +1,106 @@
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Scanner;
```
Would love to see a separation between util & io packages here again

Ramanathan0908 commented on src/test/java/TaskListTest.java:
```diff
@@ -0,0 +1,27 @@
+import java.util.Arrays;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.fail;
```
The import statements could be arranged better, i.e the junit ones, then java ones then duke ones etc.

Ramanathan0908 commented on src/main/java/Duke/util/Echo.java:
```diff
@@ -0,0 +1,17 @@
+        }
+        System.out.println("Bye. Hope to see you again soon");
+    }
+}
```
unless I am missing something, what is the point of the echo class?

Ramanathan0908 reviewed (commented):
Great job overall!!! Just some minor nits to fix!

quannam0124 committed:
Create GUI using JavaFX

Duke uses Text-based interface to communicate with users

Creating GUI using JavaFX allows a better user experience

Let's refactor Duke to return replies in String, and use JavaFX
to display them to users

quannam0124 committed:
Add javadocs

quannam0124 committed:
Add assertions

Duke Manager class does not have an assertion

Let's add an assertion to notify if the response from Duke is empty

quannam0124 committed:
Add assertions

Duke Manager class does not have an assertion

Let's add an assertion to notify if the response from Duke is empty

quannam0124 committed:
Add print command for util.UI

quannam0124 committed:
Extract command functions

Duke uses a DukeHandler to process input and return the output

Let's extract the parser and command processor to achieve further
OOP

quannam0124 committed:
Improve Code Quality

Duke uses too long methods

Let's extract out some common methods to increase readability

quannam0124 committed:
Merge pull request #2 from quannam0124/branch-A-CodeQuality

Improve code quality

quannam0124 committed:
Merge branch 'master' into branch-A-Assertions

quannam0124 committed:
Merge branch 'master' into branch-A-Assertions

# Conflicts:
#	src/main/java/Duke/DukeHandler.java
#	src/main/java/duke/DukeManager.java

quannam0124 committed:
Merge pull request #3 from quannam0124/branch-A-Assertions

Branch a assertions

quannam0124 committed:
Add Help extension

Duke does not have a help command to show users what they can do with
the program

Let's add a help command and let users know that they can use it to see
all the commands and their formats

quannam0124 committed:
Merge pull request #4 from quannam0124/branch-C-Help

Add Help extension

quannam0124 committed:
ExitCommand: Program exit delay

Duke exits before showing the bye output

By informing the user that the program is exiting, we can reassure them
that it was not a crash

Let's add a timetask function so that the program can display a goodbye
message before exiting

quannam0124 committed:
Add TaskNotFound Exception

Duke runs does not return anything if IndexOutofBound Exception
occurred due to the user's input mark, unmark, delete's index is
larger than the task list size

Let's add TaskNotFound Exception to handle this exception and show
the user that an error has come up because of their input

quannam0124 committed:
Add User Guide, add JUnit test for Command

quannam0124 committed:
Improve GUI

Duke's font size is small and hard to read. The profile photos are
square

Let's increase the size of the font and crop the profile images so
users will have a more enjoyable experience using Duke

quannam0124 committed:
Add Exception handling

Duke does not handle exception when user inputs the incorrect date
format

Let's add DukeDateException to handle this error

quannam0124 committed:
Add Javadocs

quannam0124 committed:
Add User Guide and screenshot

quannam0124 committed:
Update User guide

quannam0124 committed:
Update User Guide

quannam0124 committed:
Set title of stage to be Duke Manager

quannam0124 committed:
Enable assertions

quannam0124 committed:
Update readme.md
